### ISSUE TITLE
Prefer yield without arguments to block.call where possible

### ISSUE THREAD

**Issue by nateberkopec** [dgopstein/deep_enumerable#5]
Unfortunately, [creating blocks in MRI is still very expensive](https://www.omniref.com/ruby/2.2.0/symbols/Proc/yield?#annotation=4087638&line=711&hn=1), and you seem to pass around blocks quite a bit in this library. I could see this lib being used in some critical path/performance sensitive areas as well, so I think performance in this area is important.

I didn't take a deep enough dive in this to see where this would be possible (for example, [this bit in deep_diff](https://github.com/dgopstein/deep_enumerable/blob/master/lib/deep_enumerable.rb#L25) could probably use yield and block_given?) Otherwise, I think the library is extremely well written!

**Comment by dgopstein** [comment]
Oof. I had no idea about the performance implications of block arguments. I think you're right that it makes sense to try to address this if possible. Since this bottleneck happens in such a nuanced way it's probably also worth putting in a few benchmark tests to make sure whatever fix is applied actually fixes the issue.

Thanks for spotting and reporting this!

**Comment by nateberkopec** [comment]
Agreed on tests. [Minitest::Benchmark](https://github.com/seattlerb/minitest/blob/master/lib/minitest/benchmark.rb) is a great little tool for this.
